units/src/main/java/net/consensys/cava/units/bigints/UInt64Value.java

```diff
@@ -0,0 +1,383 @@
+/*
+ * Copyright 2018 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package net.consensys.cava.units.bigints;
+
+
+import net.consensys.cava.bytes.Bytes;
+
+import java.math.BigInteger;
+
+/**
+ * Represents a 64-bit (8 bytes) unsigned integer value.
+ *
+ * <p>
+ * A {@link UInt64Value} is an unsigned integer value stored with 8 bytes, so whose value can range between 0 and
+ * 2^64-1.
+ *
+ * <p>
+ * This interface defines operations for value types with a 64-bit precision range. The methods provided by this
+ * interface take parameters of the same type (and also {@code long}. This provides type safety by ensuring calculations
+ * cannot mix different {@code UInt64Value} types.
+ *
+ * <p>
+ * Where only a pure numerical 64-bit value is required, {@link UInt64} should be used.
+ *
+ * <p>
+ * It is strongly advised to extend {@link BaseUInt64Value} rather than implementing this interface directly. Doing so
+ * provides type safety in that quantities of different units cannot be mixed accidentally.
+ *
+ * @param <T> The concrete type of the value.
+ */
+public interface UInt64Value<T extends UInt64Value<T>> extends Comparable<T> {
+
+  /**
+   * @return True if this is the value 0.
+   */
+  default boolean isZero() {
+    return toBytes().isZero();
+  }
+
+  /**
+   * Returns a value that is {@code (this + value)}.
+   *
+   * @param value The amount to be added to this value.
+   * @return {@code this + value}
+   */
+  T add(T value);
+
+  /**
+   * Returns a value that is {@code (this + value)}.
+   *
+   * @param value The amount to be added to this value.
+   * @return {@code this + value}
+   * @throws ArithmeticException if the result of the addition overflows
+   */
+  default T tryAdd(T value) {
+    T result = add(value);
+    if (this.compareTo(result) > 0) {
+      throw new ArithmeticException("Value overflowed");
+    }
+    return result;
+  }
+
+  /**
+   * Returns a value that is {@code (this + value)}.
+   *
+   * @param value The amount to be added to this value.
+   * @return {@code this + value}
+   */
+  T add(long value);
+
+  /**
+   * Returns a value that is {@code (this + value)}.
+   *
+   * @param value The amount to be added to this value.
+   * @return {@code this + value}
+   * @throws ArithmeticException if the result of the addition overflows
+   */
+  default T tryAdd(long value) {
+    T result = add(value);
+    if (this.compareTo(result) > 0) {
```

Comment:
This check is insufficient to detect overflows that wrap further than the original value.